# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webcodecs/VideoDecoder.cpp
# Commit: d43f0c82f214
# Full Hash: d43f0c82f2148e9e4353afa5a4fd9fa4075a260f
# Author: Paul Adenot <paul@paul.cx>
# Date: 2023-12-23 09:52:03
# Regressor Bug: 1749047
# File Overlap Count: 1
# Description:
#   Bug 1749047 - Guess more colorspace information when missing, based on the information present. r=pehrsons
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D196470
# ==============================================================================

diff -r c781f30e5605 -r d43f0c82f214 dom/media/webcodecs/VideoDecoder.cpp
--- a/dom/media/webcodecs/VideoDecoder.cpp	Fri Dec 22 21:39:16 2023 +0000
+++ b/dom/media/webcodecs/VideoDecoder.cpp	Fri Dec 22 21:39:16 2023 +0000
@@ -154,7 +154,7 @@
       OptionalToMaybe(aConfig.mOptimizeForLatency)));
 }
 
-nsString VideoDecoderConfigInternal::ToString() {
+nsString VideoDecoderConfigInternal::ToString() const {
   nsString rv;
 
   rv.Append(mCodec);
@@ -381,15 +381,67 @@
   colorSpace.mFullRange = Some(ToFullRange(aData->mColorRange));
   if (Maybe<VideoMatrixCoefficients> m =
           ToMatrixCoefficients(aData->mYUVColorSpace)) {
-    colorSpace.mMatrix = Some(*m);
+    colorSpace.mMatrix = ToMatrixCoefficients(aData->mYUVColorSpace);
+    colorSpace.mPrimaries = ToPrimaries(aData->mColorPrimaries);
   }
-  if (Maybe<VideoColorPrimaries> p = ToPrimaries(aData->mColorPrimaries)) {
-    colorSpace.mPrimaries = Some(*p);
+  if (!colorSpace.mPrimaries) {
+    LOG("Missing primaries, guessing from colorspace");
+    // Make an educated guess based on the coefficients.
+    colorSpace.mPrimaries = colorSpace.mMatrix.map([](const auto& aMatrix) {
+      switch (aMatrix) {
+        case VideoMatrixCoefficients::EndGuard_:
+          MOZ_CRASH("This should not happen");
+        case VideoMatrixCoefficients::Bt2020_ncl:
+          return VideoColorPrimaries::Bt2020;
+        case VideoMatrixCoefficients::Rgb:
+        case VideoMatrixCoefficients::Bt470bg:
+        case VideoMatrixCoefficients::Smpte170m:
+          LOGW(
+              "Warning: Falling back to BT709 when attempting to determine the "
+              "primaries function of a YCbCr buffer");
+          [[fallthrough]];
+        case VideoMatrixCoefficients::Bt709:
+          return VideoColorPrimaries::Bt709;
+      }
+      MOZ_ASSERT_UNREACHABLE("Unexpected matrix coefficients");
+      LOGW(
+          "Warning: Falling back to BT709 due to unexpected matrix "
+          "coefficients "
+          "when attempting to determine the primaries function of a YCbCr "
+          "buffer");
+      return VideoColorPrimaries::Bt709;
+    });
   }
+
   if (Maybe<VideoTransferCharacteristics> c =
           ToTransferCharacteristics(aData->mTransferFunction)) {
     colorSpace.mTransfer = Some(*c);
   }
+  if (!colorSpace.mTransfer) {
+    LOG("Missing transfer characteristics, guessing from colorspace");
+    colorSpace.mTransfer = Some(([&] {
+      switch (aData->mYUVColorSpace) {
+        case gfx::YUVColorSpace::Identity:
+          return VideoTransferCharacteristics::Iec61966_2_1;
+        case gfx::YUVColorSpace::BT2020:
+          return VideoTransferCharacteristics::Pq;
+        case gfx::YUVColorSpace::BT601:
+          LOGW(
+              "Warning: Falling back to BT709 when attempting to determine the "
+              "transfer function of a MacIOSurface");
+          [[fallthrough]];
+        case gfx::YUVColorSpace::BT709:
+          return VideoTransferCharacteristics::Bt709;
+      }
+      MOZ_ASSERT_UNREACHABLE("Unexpected color space");
+      LOGW(
+          "Warning: Falling back to BT709 due to unexpected color space "
+          "when attempting to determine the transfer function of a "
+          "MacIOSurface");
+      return VideoTransferCharacteristics::Bt709;
+    })());
+  }
+
   return colorSpace;
 }
 
